nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/processors/slack/PutSlack.java

```diff
@@ -0,0 +1,245 @@
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.apache.nifi.stream.io.DataOutputStream;
+
+import javax.json.Json;
+import javax.json.JsonObject;
+import javax.json.JsonObjectBuilder;
+import javax.json.JsonWriter;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+@Tags({"put", "slack", "notify"})
+@CapabilityDescription("Sends a message to your team on slack.com")
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+public class PutSlack extends AbstractProcessor {
+
+    public static final PropertyDescriptor WEBHOOK_URL = new PropertyDescriptor
+            .Builder()
+            .name("webhook-url")
+            .displayName("Webhook URL")
+            .description("The POST URL provided by Slack to send messages into a channel.")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .addValidator(StandardValidators.URL_VALIDATOR)
+            .sensitive(true)
+            .build();
+
+    public static final PropertyDescriptor WEBHOOK_TEXT = new PropertyDescriptor
+            .Builder()
+            .name("webhook-text")
+            .displayName("Webhook Text")
+            .description("The text sent in the webhook message")
+            .required(true)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor CHANNEL = new PropertyDescriptor
+            .Builder()
+            .name("channel")
+            .displayName("Channel")
+            .description("A public channel using #channel or direct message using @username. If not specified, " +
+                    "the default webhook channel as specified in Slack's Incoming Webhooks web interface is used.")
+            .required(false)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor USERNAME = new PropertyDescriptor
+            .Builder()
+            .name("username")
+            .displayName("Username")
+            .description("The displayed Slack username")
+            .required(false)
+            .expressionLanguageSupported(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor ICON_URL = new PropertyDescriptor
+            .Builder()
+            .name("icon-url")
+            .displayName("Icon URL")
+            .description("Icon URL to be used for the message")
```

Comment:
I found out the hard way that if you use Icon URL, you can't use Username (or else the Icon URL is ignored). Also if you use Icon URL and Emoji, it is the Emoji that is displayed (so Icon URL is ignored again). I know that's Slack's behavior not this processor's, but would be helpful to have that in the description(s) for the properties, as well as a customValidate() method that checks these. I know the actual values can't be validated as they support Expression Language, but certainly their presence or absence can be checked.